src/Compilers/Core/Portable/Syntax/AbstractWarningStateMap.cs

```diff
@@ -87,5 +87,13 @@ public int CompareTo(WarningStateMapEntry other)
                 return this.Position - other.Position;
             }
         }
+
+        protected class WarningStateMapEntryComparer : System.Collections.IComparer
+        {
+            public int Compare(object x, object y)
+            {
+                return ((WarningStateMapEntry)x).CompareTo((WarningStateMapEntry)y);
```

Comment:
The `IComparer` interface implementations should throw `ArgumentException` when given values of improper types.  This will throw an `InvalidCastException`